// react-figma/renderers/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var rectangle_1 = require("./rectangle");
exports.rectangle = rectangle_1.rectangle;
var text_1 = require("./text");
exports.text = text_1.text;
var page_1 = require("./page");
exports.page = page_1.page;
var frame_1 = require("./frame");
exports.frame = frame_1.frame;
var component_1 = require("./component");
exports.component = component_1.component;
var componentset_1 = require("./componentset");
exports.componentset = componentset_1.componentset;
var instance_1 = require("./instance");
exports.instance = instance_1.instance;
var star_1 = require("./star");
exports.star = star_1.star;
var vector_1 = require("./vector");
exports.vector = vector_1.vector;
var line_1 = require("./line");
exports.line = line_1.line;
var group_1 = require("./group");
exports.group = group_1.group;
var ellipse_1 = require("./ellipse");
exports.ellipse = ellipse_1.ellipse;
var svg_1 = require("./svg");
exports.svg = svg_1.svg;
var slice_1 = require("./slice");
exports.slice = slice_1.slice;
