// react-figma/helpers/StyleSheet.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
var absoluteFill = {
    position: 'absolute',
    left: 0,
    top: 0,
    width: '100%',
    height: '100%'
};
var StyleSheet = /** @class */ (function () {
    function StyleSheet() {
    }
    StyleSheet.create = function (styles) {
        return styles;
    };
    StyleSheet.flatten = function (styles) {
        if (Array.isArray(styles)) {
            return styles.reduce(function (acc, item) { return (__assign({}, acc, item)); }, {});
        }
        else {
            return styles || {};
        }
    };
    StyleSheet.resolve = function (value) {
        return value;
    };
    StyleSheet.compose = function (style1, style2) {
        return __assign({}, (style1 || {}), (style2 || {}));
    };
    StyleSheet.hairlineWidth = 1;
    StyleSheet.absoluteFill = absoluteFill;
    StyleSheet.absoluteFillObject = absoluteFill;
    return StyleSheet;
}());
exports.StyleSheet = StyleSheet;
