// react-figma/styleTransformers/transformAutoLayoutToYoga.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
var layoutModeToFlexDirection = {
    HORIZONTAL: 'row',
    VERTICAL: 'column'
};
var layoutAlignToAlignSelf = {
    MIN: 'flex-start',
    CENTER: 'center',
    MAX: 'flex-end',
    STRETCH: 'stretch'
};
var primaryAxisAlignItemsToJustifyContent = {
    MIN: 'flex-start',
    MAX: 'flex-end',
    CENTER: 'center',
    SPACE_BETWEEN: 'space-between'
};
var counterAxisAlignItemsToAlignItems = {
    MIN: 'flex-start',
    MAX: 'flex-end',
    CENTER: 'center'
};
exports.transformAutoLayoutToYoga = function (props) {
    return __assign({}, (props.layoutMode && layoutModeToFlexDirection[props.layoutMode]
        ? {
            flexDirection: layoutModeToFlexDirection[props.layoutMode]
        }
        : {}), (props.layoutAlign && layoutAlignToAlignSelf[props.layoutAlign]
        ? {
            alignSelf: layoutAlignToAlignSelf[props.layoutAlign]
        }
        : {}), (props.layoutGrow
        ? {
            flexGrow: props.layoutGrow
        }
        : {}), (props.horizontalPadding
        ? {
            paddingLeft: props.horizontalPadding,
            paddingRight: props.horizontalPadding
        }
        : {}), (props.verticalPadding
        ? {
            paddingTop: props.verticalPadding,
            paddingBottom: props.verticalPadding
        }
        : {}), (props.paddingLeft
        ? {
            paddingLeft: props.paddingLeft
        }
        : {}), (props.paddingRight
        ? {
            paddingRight: props.paddingRight
        }
        : {}), (props.paddingTop
        ? {
            paddingTop: props.paddingTop
        }
        : {}), (props.paddingBottom
        ? {
            paddingBottom: props.paddingBottom
        }
        : {}), (props.primaryAxisAlignItems && primaryAxisAlignItemsToJustifyContent[props.primaryAxisAlignItems]
        ? {
            justifyContent: primaryAxisAlignItemsToJustifyContent[props.primaryAxisAlignItems]
        }
        : {}), (props.counterAxisAlignItems && counterAxisAlignItemsToAlignItems[props.counterAxisAlignItems]
        ? {
            alignItems: counterAxisAlignItemsToAlignItems[props.counterAxisAlignItems]
        }
        : {}));
};
