// react-figma/styleTransformers/transformDimension.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.transformDimension = function (value) {
    if (value === 'auto') {
        return { type: 'auto' };
    }
    return {
        value: typeof value == 'number' ? value : parseFloat(value),
        type: typeof value == 'number' || value.indexOf('%') < 0 ? 'px' : 'percentage'
    };
};
exports.transformDimensionMapper = function (value) {
    var result = exports.transformDimension(value);
    var returned;
    var holder = {
        px: function (f) {
            if (result.type === 'px') {
                returned = f(result.value);
            }
            return holder;
        },
        percentage: function (f) {
            if (result.type === 'percentage') {
                returned = f(result.value);
            }
            return holder;
        },
        auto: function (f) {
            if (result.type === 'auto') {
                returned = f();
            }
            return holder;
        },
        value: function () {
            return returned;
        }
    };
    return holder;
};
