// react-figma/rn/Dimensions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var DEFAULT_SCALED_SIZE = {
    fontScale: 1,
    height: Number.MAX_SAFE_INTEGER,
    scale: 1,
    width: Number.MAX_SAFE_INTEGER
};
var DimensionsImpl = /** @class */ (function () {
    function DimensionsImpl() {
        this.dims = {
            window: DEFAULT_SCALED_SIZE,
            screen: DEFAULT_SCALED_SIZE
        };
        this.changeListeners = [];
    }
    DimensionsImpl.prototype.addEventListener = function (type, handler) {
        this.changeListeners = this.changeListeners.filter(function (l) { return l !== handler; }).concat([handler]);
    };
    DimensionsImpl.prototype.get = function (dim) {
        return this.dims[dim];
    };
    DimensionsImpl.prototype.removeEventListener = function (type, handler) {
        this.changeListeners = this.changeListeners.filter(function (l) { return l !== handler; }).slice();
    };
    DimensionsImpl.prototype.set = function (dims) {
        var _this = this;
        this.dims = dims;
        this.changeListeners.forEach(function (listener) { return listener(_this.dims); });
    };
    return DimensionsImpl;
}());
exports.Dimensions = new DimensionsImpl();
