// react-figma/yoga/yogaSerializer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var yoga = require("yoga-layout-prebuilt");
var yogaEdge = {
    left: yoga.EDGE_LEFT,
    top: yoga.EDGE_TOP,
    right: yoga.EDGE_RIGHT,
    bottom: yoga.EDGE_BOTTOM,
    start: yoga.EDGE_START,
    end: yoga.EDGE_END,
    horizontal: yoga.EDGE_HORIZONTAL,
    vertical: yoga.EDGE_VERTICAL,
    all: yoga.EDGE_ALL
};
var config = {
    alignContent: 'getAlignContent',
    alignItems: 'getAlignItems',
    alignSelf: 'getAlignSelf',
    aspectRatio: 'getAspectRatio',
    border: ['getBorder', yogaEdge],
    display: 'getDisplay',
    flexBasis: 'getFlexBasis',
    flexDirection: 'getFlexDirection',
    flexGrow: 'getFlexGrow',
    flexShrink: 'getFlexShrink',
    flexWrap: 'getFlexWrap',
    height: 'getHeight',
    justifyContent: 'getJustifyContent',
    margin: ['getMargin', yogaEdge],
    maxHeight: 'getMaxHeight',
    maxWidth: 'getMaxWidth',
    minHeight: 'getMinHeight',
    minWidth: 'getMinWidth',
    overflow: 'getOverflow',
    padding: ['getPadding', yogaEdge],
    position: ['getPosition', yogaEdge],
    positionType: 'getPositionType',
    width: 'getWidth'
};
exports.serializeYogaNodeStyle = function (yogaNode) {
    var result = {};
    Object.keys(config).forEach(function (prop) {
        var getter = config[prop];
        if (typeof getter === 'string') {
            result[prop] = yogaNode[getter]();
        }
        else {
            var f_1 = getter[0];
            var args_1 = getter[1];
            result[prop] = {};
            Object.keys(args_1).forEach(function (argKey) {
                result[prop][argKey] = yogaNode[f_1](args_1[argKey]);
            });
        }
    });
    return result;
};
